wp-includes/components/class-wp-service-worker-caching-routes.php

```diff
@@ -66,22 +80,64 @@ class WP_Service_Worker_Caching_Routes implements WP_Service_Worker_Registry {
 	 * @since 0.2
 	 *
 	 * @param string $route Route regular expression, without delimiters.
-	 * @param array  $args  {
+	 * @param array  $args {
 	 *     Additional route arguments.
 	 *
-	 *     @type string $strategy   Required. Strategy, can be WP_Service_Worker_Caching_Routes::STRATEGY_STALE_WHILE_REVALIDATE, WP_Service_Worker_Caching_Routes::STRATEGY_CACHE_FIRST,
-	 *                              WP_Service_Worker_Caching_Routes::STRATEGY_NETWORK_FIRST, WP_Service_Worker_Caching_Routes::STRATEGY_CACHE_ONLY,
-	 *                              WP_Service_Worker_Caching_Routes::STRATEGY_NETWORK_ONLY.
-	 *     @type string $cache_name Name to use for the cache.
-	 *     @type array  $plugins    Array of plugins with configuration. The key of each plugin in the array must match the plugin's name.
-	 *                              See https://developers.google.com/web/tools/workbox/guides/using-plugins#workbox_plugins.
+	 *     @type string     $strategy           Strategy, can be WP_Service_Worker_Caching_Routes::STRATEGY_STALE_WHILE_REVALIDATE,
+	 *                                          WP_Service_Worker_Caching_Routes::STRATEGY_CACHE_FIRST, WP_Service_Worker_Caching_Routes::STRATEGY_NETWORK_FIRST,
+	 *                                          WP_Service_Worker_Caching_Routes::STRATEGY_CACHE_ONLY, WP_Service_Worker_Caching_Routes::STRATEGY_NETWORK_ONLY.
+	 *                                          Required.
+	 *     @type string     $cache_name         Name to use for the cache.
+	 *     @type array|null $expiration         Expiration plugin configuration. See <https://developers.google.com/web/tools/workbox/reference-docs/latest/module-workbox-expiration.ExpirationPlugin>.
+	 *     @type array|null $broadcast_update   Broadcast update plugin configuration. See <https://developers.google.com/web/tools/workbox/reference-docs/latest/module-workbox-broadcast-update.BroadcastUpdatePlugin>.
+	 *     @type array|null $cacheable_response Cacheable response plugin configuration. See <https://developers.google.com/web/tools/workbox/reference-docs/latest/module-workbox-cacheable-response.CacheableResponsePlugin>.
+	 *     @type array|null $background_sync    Background sync plugin configuration. See <https://developers.google.com/web/tools/workbox/reference-docs/latest/module-workbox-background-sync.BackgroundSyncPlugin>.
+	 *     @type array|null $plugins            Deprecated. Array of plugins with configuration. The key of each plugin in the array must match the plugin's name.
+	 *                                          This is deprecated in favor of defining the plugins in the top-level.
+	 *                                          See <https://developers.google.com/web/tools/workbox/guides/using-plugins#workbox_plugins>.
+	 * @return bool Whether the registration was successful.
 	 * }
 	 */
 	public function register( $route, $args = array() ) {
```

Comment:
I think the changes made here indicate that it might be more intuitive to make `route` part of `$args` here too and have that as the only parameter. The `strategy` argument is required anyway, so I think this array overall should be required.

We could change the method signature to something like `register( $args, $deprecated = array() )`, and then if `$args` is a string, assume that it is actually `$route` and then `$deprecated` would be `$args`.

Reply:
I actually went down that “route” in the course of this PR, but then I decided against it. The reason I decided to keep two arguments is that it keeps the method signature consistent with `WP_Service_Worker_Precaching_Routes::register()`. Both then take a URL/route and then args. In both cases, the args can be a string. For precaching, the `$args` can be a string as a shortcut to be the `revision`. For `WP_Service_Worker_Caching_Routes::register()`, the `$args` as a string is a shortcut to provide the `strategy`. 

It's kinda handy to be able to do:

```php
$scripts->caching_routes()->register( '/foo', WP_Service_Worker_Caching_Routes::STRATEGY_NETWORK_FIRST ); 
```

Rather than:

```php
$scripts->caching_routes()->register( '/foo', array(
    'strategy' => WP_Service_Worker_Caching_Routes::STRATEGY_NETWORK_FIRST,
) ); 
```

So that's the intention here.

Reply:
Hmm that's a fair point. However, then I'd suggest to give both of these methods three parameters: The URL/route and the strategy/revision are both required, so I think it's counterintuitive to have these in `$args`. IMO `$args` should always be an optional thing for more "advanced" customizations.

Reply:
So you're saying for caching routes it should be:

```php
$scripts->caching_routes()->register( 
    '/foo', 
    WP_Service_Worker_Caching_Routes::STRATEGY_NETWORK_FIRST,
    array( /* Additional optional args. */ )
); 
```

And for precaching routes (which is already supported):

```php
$scripts->precaching_routes()->register( 
    '/foo', 
    $revision
); 
```

Oh, but actually `revision` is _not_ required in the [precache list](https://developers.google.com/web/tools/workbox/modules/workbox-precaching#explanation_of_the_precache_list). But it's the only argument that precaching currently supports. There are no other arguments. So this is fine too:

```php
$scripts->precaching_routes()->register( '/foo' ); 
```

This isn't the case for caching routes, where the strategy must be provided. So in that case, `register` for caching routes would have 3 arguments but `register` for precaching routes would have only 2?

Reply:
Yes, the signatures you put as examples look good to me. Regarding `$revision` not being required, I think that's okay, in this case we can make it optional. Now, overall I think optional arguments make sense to go into an `$args` array, but if `$revision` is the only one, I don't think that would be needed here - so I suggest to make it an optional string parameter and not have `$args` for precaching routes at all.

If at some point we need additional optional arguments for precaching routes, we can transform the second parameter to an optional `$args` and keep backward-compatibility by still supporting a string (and then assuming that is the revision).

Reply:
Ah, problem: a method signature params of `( $route, $strategy, $args = array() )` will violate the implementation of `WP_Service_Worker_Registry::register` interface, which has `( $handle, $args = array() )`. Ditch the interface?

Reply:
@westonruter Yeah, I think that makes sense. We should then remove `WP_Service_Worker_Registry_Aware` as well.